go/billing/api.py

```diff
@@ -808,28 +835,83 @@ def create_transaction_interaction(self, cursor, account_number,
-    def notification_threshold_crossed(
-            credit_balance, credit_amount, alert_credit_balance):
+    def check_and_notify_low_credit_threshold(
+            self, credit_balance, credit_amount, last_topup_balance,
+            account_number):
+        """
+        Checks the current balance percentage against all those stored within
+        the settings. Sends the notification email if it is required. Returns
+        the alert percent if email was sent, or ``None`` if no email was sent.
+
+        :param credit_balance: The current balance (after the transaction)
+        :param credit_amount: The amount of credits used in the transaction
+        :param last_topup_balance: The account credit balance at the last topup
+        :param account_number: The account number of the associated account
+        """
+        level = self.check_all_low_credit_thresholds(
+            credit_balance, credit_amount, last_topup_balance)
+        if level is not None:
+            return spawn_celery_task_via_thread(
+                create_low_credit_notification, account_number,
+                level, credit_balance)
+
+    def _get_notification_level(self, percentage):
+        """
+        Fetches the value of the notification level for the given percentage.
+
+        :param int percentage:
+            The percentage to get the notification level for
+
+        :return:
+            An int representing the current notification level.
+        """
+        if not self._notification_mapping:
+            return None
+
+        minimum = self._notification_mapping[0]
+        if percentage < minimum:
+            return minimum
+        if percentage > self._notification_mapping[-1]:
+            return None
+        return self._notification_mapping[percentage - minimum]
+
+    def check_all_low_credit_thresholds(
+            self, credit_balance, credit_amount, last_topup_balance):
         """
-        Given the current balance (afther the transaction) ``credit_balance``,
-        the transaction amount ``credit_amount``, and the alert threshold
-        ``alert_credit_balance``, will return ``True`` if the transaction
-        caused the alert threshold to be crossed, and false if not.
+        Checks the current balance percentage against all those stored within
+        the settings.
+
+        :param credit_balance:
+            The current balance (after the transaction)
+        :param credit_amount:
+            The amount of credits used in the transaction
+        :param last_topup_balance:
+            The account credit balance at the last topup
+
+        :return:
+            A :class:`Decimal` percentage for the alert threshold crossed
+            or ``None`` if no threshold was crossed.
         """
-        return (
-            credit_balance <= alert_credit_balance <
-            credit_amount + credit_balance)
+        def ceil_percent(n):
+            return int(math.ceil(n * 100 / last_topup_balance))
```

Comment:
We need to handle and test the case where ``last_topup_balance`` is ``0`` (or perhaps ``None``) because once we deploy this there will be a lot of accounts where the topup balance hasn't been set yet.